packages/notifier/src/notifier.js

```diff
@@ -2,56 +2,68 @@ import { producePromise } from '@agoric/produce-promise';
+ * If the handle argument is omitted or differs from the current handle, the
+ * current record is returned. If the handle corresponds to the current state,
+ * a promise is returned. After the next state change, The promise will resolve
+ * to the then-current value of the record.
  *
  * updateState(state) sets a new state and resolves the outstanding promise.
- * This method should be closely held, as anyone with access to it can send
- * updates.
+ * resolve(finalState) sets the new state, sends a final update, and freezes the
+ * updater. The updater object should be closely held, as anyone with access to
+ * it can provide updates.
  */
-export const makeNotifier = () => {
-  let updateHandle;
-  let currentPromiseRec;
-  let currentState;
-  let currentResponse;
-
-  const updateResponse = (closeStream = false) => {
-    updateHandle = closeStream ? undefined : {};
-    currentPromiseRec = producePromise();
-    currentResponse = harden({ currentState, updateHandle });
-    if (closeStream) {
-      currentPromiseRec.resolve(currentResponse);
-    }
-  };
-  updateResponse();
-
-  const notifier = harden({
-    getUpdateSince: (lastUpdateHandle = undefined) => {
-      if (lastUpdateHandle === updateHandle) {
-        return currentPromiseRec.promise;
-      }
-
-      return currentResponse;
-    },
-    updateState: (state, closeStream = false) => {
-      if (!updateHandle) {
-        throw new Error('Cannot update state after close.');
-      }
-
-      currentState = state;
-      const savedPromise = currentPromiseRec;
-      updateResponse(closeStream);
-      savedPromise.resolve(currentResponse);
-    },
+export const produceNotifier = () => {
+  let currentPromiseRec = producePromise();
+  let currentResponse = harden({
+    value: undefined,
+    updateHandle: {},
+    done: false,
   });
-  return notifier;
+
+  function getUpdateSince(updateHandle = undefined) {
+    if (updateHandle === currentResponse.updateHandle) {
+      return currentPromiseRec.promise;
+    }
+
+    return currentResponse;
+  }
+
+  function resolve(finalState) {
+    if (!currentResponse.updateHandle) {
+      throw new Error('Cannot resolve again.');
+    }
+
+    currentResponse = harden({
+      value: finalState,
+      updateHandle: undefined,
+      done: true,
+    });
+    currentPromiseRec.resolve(currentResponse);
+  }
+
+  function updateState(state) {
+    if (!currentResponse.updateHandle) {
+      throw new Error('Cannot update state after close.');
+    }
+
+    currentResponse = harden({ value: state, updateHandle: {}, done: false });
+    currentPromiseRec.resolve(currentResponse);
+    currentPromiseRec = producePromise();
+  }
+
+  // notifier facet is separate so it can be handed out loosely while updater is
+  // tightly held
+  const notifier = harden({ getUpdateSince });
+  const updater = harden({ resolve, updateState });
+  return harden({ updater, notifier });
```

Comment:
Please swap order so it reads
```js
{ updateState, resolve }
```
so temporal chronology reads left-to-right.

Please swap order so it reads
```js
{ notifier, updater }
```
so that it parallels
```js
{ promise, resolve }
```
where the query facet precedes the update facet.

Reply:
done